sale_order_lot_selection/models/sale.py

```diff
@@ -0,0 +1,79 @@
+from odoo import _, api, fields, models
+from odoo.exceptions import UserError
+
+
+class SaleOrderLine(models.Model):
+    _inherit = "sale.order.line"
+
+    lot_id = fields.Many2one("stock.production.lot", "Lot", copy=False)
+
+    @api.onchange("product_id")
+    def product_id_change(self):
+        super(SaleOrderLine, self).product_id_change()
+        self.lot_id = False
+
+    @api.onchange("product_id")
+    def _onchange_product_id_set_lot_domain(self):
+        available_lot_ids = []
+        if self.order_id.warehouse_id and self.product_id:
+            location = self.order_id.warehouse_id.lot_stock_id
+            quants = self.env["stock.quant"].read_group(
+                [
+                    ("product_id", "=", self.product_id.id),
+                    ("location_id", "child_of", location.id),
+                    ("quantity", ">", 0),
+                    ("lot_id", "!=", False),
+                ],
+                ["lot_id"],
+                "lot_id",
+            )
+            available_lot_ids = [quant["lot_id"][0] for quant in quants]
+        self.lot_id = False
+        return {"domain": {"lot_id": [("id", "in", available_lot_ids)]}}
+
+
+class SaleOrder(models.Model):
+    _inherit = "sale.order"
+
+    @api.model
+    def get_move_from_line(self, line):
+        move = self.env["stock.move"]
+        # i create this counter to check lot's univocity on move line
+        lot_count = 0
+        for p in line.order_id.picking_ids:
```

Comment:
Is it possible to avoid loop in loop? I think it's possible to have only 1.

Reply:
Test case are failing, if I try to implement logic to avoid loop. So it would be best, if you propose your changes.

Reply:
```
for p in line.order_id.picking_ids:
    for m in p.move_lines:
        ...
```
Couldn't be changed into:
```
for m in line.order_id.picking_ids.mapped("move_lines"):
    ...
```
As `p` is not used at all?